### ISSUE TITLE
Problems with Tmux 2.4

### ISSUE THREAD

**Issue by thalesmello** [tmux-plugins/tmux-copycat#109]
Since tmux 2.4 (currently in development), the way vi key bindings for copy-mode and the dynamics for send-keys have changed.

Therefore, this plugin currently doesn't work.

In case there is anyone interested, I have a [fork](https://github.com/thalesmello/tmux-copycat) that I managed to make it work for copy-mode in vi key bindings (emacs would need a bit more of work).

I won't open a pull request as that would break compatibility with previous versions.

**Comment by floscr** [comment]
Thank you, this was driving me nuts! 👍

**Comment by dominikduda** [comment]
@thalesmello U seems to be maintaining this plugin. Coul you please take a look at my question? No one will respond to it here probably.

https://github.com/tmux-plugins/tmux-copycat/issues/111

Thanks!

**Comment by oblitum** [comment]
Hi, I think I've been hit by this now, as well as any other ArchLinux user, please check https://github.com/tmux-plugins/tmux-yank/issues/76. The plugin sadly simply stopped working.

**Comment by thalesmello** [comment]
@oblitum Thank you for pointing the issue with tmux-yank. I wasn't aware of it.

It wouldn't take much effort to adjust my fork to make it work with previous versions of Tmux.

I don't have time to do it now. I'll open a pull request as soon as I get a chance to.

**Comment by oblitum** [comment]
@thalesmello thanks, I'm using your fork meanwhile.

**Comment by oblitum** [comment]
@thalesmello I've figured out your patch is not working for me in the end:

- with tmux-yank working (https://github.com/tmux-plugins/tmux-yank/issues/76 applied)
- check that simply select and yank is working fine
- now try `prefix + ctrl-u` or any, to select, or even not select anything at all
- tmux-copycat selection will work
- yank of such selection appears to work since selection vanishes
- now tmux-yank is broken, when pasting I see it didn't yank the copycat selection, it is still with the original.
- I can't yank anything anymore, I have to reload `.tmux.conf` for tmux-yank to work again.

**Comment by thalesmello** [comment]
@oblitum Here is why.

Copycat tries to figure out which keys exit copy mode in order to override the original hooks to clear up copycat specific keybindings when you exit copy mode. The problem is that it assumes that any keybinding that has the `cancel` keyword exits copy-mode.

All the while, the tmux-yank plugin registers keybindings using `copypipe-and-cancel` keybinding, which also exits copy mode.

So here is what happens:

1. tmux-yank registers `copypipe-and-cancel` keybindings to copy text.
2. tmux-copycat overrides these shortcuts with special hooks to clear up copycat keybindings when the user wants to leave copy mode. Note that this also overrides the original yank behavior.
3. After tmux-copycat exists copy mode, it replaces the original keybindings by `cancel` shortcuts. And that's why tmux-yank stops working after that.

So, in order to solve this issue, copycat has to be smarter in order to know the original keybindings, all of that in order to be able to correctly exit copy mode and clear copycat's keybindings (without losing the yank capability) and also be able to restore the original keybindings.

That would be quite a bit more of work, which I didn't have to do because I simply use `copypipe` (without tmux-yank) and that doesn't conflict with copycat's keybindings.

I can try to take a look at it when I have time, but I have no idea when I will take a look at it.

In case you want to use by approach, the relevant keybindings are as follows:

```tmux
bind-key -Tcopy-mode-vi y send -X copy-pipe "reattach-to-user-namespace pbcopy"\; send -X clear-selection
bind-key -Tcopy-mode-vi Enter send -X copy-pipe "reattach-to-user-namespace pbcopy"\; send -X clear-selection
```

**Comment by oblitum** [comment]
I'm freezing to 2.3 for now... too many issues, all plugins (copycat, yank, open, ...?) suddenly stopped working because of the [breaking change in 2.4](https://github.com/tmux/tmux/commit/76d6d3641f271be1756e41494960d96714e7ee58).

**Comment by thalesmello** [comment]
I also have a fork I'm which I made open work for me. I will take a look
into fixing this plugin as soon as possible.

**Comment by davidjb** [comment]
@thalesmello Just a heads up that for me, your fork can select copycat results several characters too late (which can end up being on the next line etc).  Some quick tests imply it's the Unicode characters in my prompt (from Powerline), as any lines that look like ASCII-only are fine.  Thanks for the fork, though; partially working is heaps better than before.

**Comment by laktak** [comment]
```
bind-key -Tcopy-mode-vi Enter send -X copy-pipe "reattach-to-user-namespace pbcopy"\; send -X clear-selection
```

@thalesmello Do you know if it is possible to exit copy mode afterwards? I tried a few variations (like sending cancel after clear-selection) but then the text is not copied to the clipboard.

**Comment by thalesmello** [comment]
@davidjb Do you think you could provide for a way to reproduce this problem?

**Comment by thalesmello** [comment]
@laktak Yes, it is, but it requires changes to the code in my fork. Given the number of people in seeing this solved, I'll try to work on it this week.

**Comment by AlexKvazos** [comment]
How can I install tmux 2.3 from brew? I am getting 2.4 by default.

**Comment by AlexKvazos** [comment]
Never mind, built from source.. easier.

**Comment by dansomething** [comment]
@AlexKvazos you can use `brew switch` too. even easier. `brew switch tmux 2.3_3`

**Comment by labianchin** [comment]
@thalesmello, your fork seem very nice!

I guess it would be nice to keep compatibility with older versions, in a similar way as it as done on https://github.com/tmux-plugins/tmux-yank/pull/82/files

**Comment by thalesmello** [comment]
Hello @labianchin 
I will do it eventually. I don't have the time right now.

**Comment by arooni** [comment]
@thalesmello thanks for your efforts; I cloned your repo today, but tried to run it but I see the following
relavent info: mac os x 10.12.5 ; fish shell 2.5.0 and the following config for tmux; https://gist.github.com/c9ab1c8a6bc5bcd5a1c69fc08f50e056


1. File search doesn't seem to work ; 
2. When i do the simple \ search then try to yank a match via y or Enter it doesn't add it to my clipboard.

**Comment by thalesmello** [comment]
@arooni Are you using Vi or Emacs keybindings?

My fork is only configured to work with vim keybindings.

**Comment by arooni** [comment]
@thalesmello vi keybindings.  Sorry if that wasn't clear.

**Comment by Szero** [comment]
Hi, I'm using tmux 2.5 and I noticed that this plugin unbinds all your bindings that exit from copy mode (all that contain `cancel` keyword). I'm using this keybind `bind-key -T copy-mode-vi 'y' send -X copy-pipe-and-cancel "xsel -ib"` to save copy mode selection into system's clipboard selection. I wrote simple patch to prevent this behavior and I didn't notice any negative effects caused by it. I specified `y` and `Y` key there because there are the ones that are used in the `tmux-yank` plugin. 
https://gist.github.com/Szero/dd231150f47a70aa47e3f87cb14b5fef

**Comment by oblitum** [comment]
On ArchLinux now I can't freeze on tmux 2.3 anymore because dependencies (libevent) got newer, so I'm forced to upgrade to 2.4/2.5. Sadly then, the triad yank/copycat/open still is broken because copycat is still not fixed :(

**Comment by thalesmello** [comment]
I'm having my finals this week. I want to fix this issue after Sunday.

**Comment by sflip** [comment]
Thanks to the hint of @thalesmello I could fix the issue of tmux-copycat breaking the tmux-yank functionality. I did that in yet another [fork](https://github.com/sflip/tmux-copycat).

**Comment by maximbaz** [comment]
Hey @sflip, thanks for the fix, it works _almost_ well :) One issue I found:

1. `$ echo "one\ntwo" to have some text content in terminal
2. <kbd>Prefix, /</kbd> to start search, type "one" and press <kbd>Enter</kbd>
3. "one" was correctly found, now press <kbd>Y</kbd> to copy it in the prompt
4. Now try to search "two" using the same sequence, <kbd>Prefix, /</kbd> to start search, type "two" and press <kbd>Enter</kbd>

Expected: you found "two"
Actual: the next occurrence of "one" is found, i.e. tmux keeps searching for "one", not for "two".

**Comment by bruno-** [comment]
Closing this PR as it was addressed via #119
